src/daft-catalog/src/impls/memory.rs

```diff
@@ -0,0 +1,368 @@
+
+        if self.tables.read().unwrap().contains_key(&namespace) {
+            return Err(CatalogError::obj_already_exists("namespace", ident));
+        }
+
+        self.tables
+            .write()
+            .unwrap()
+            .insert(namespace, IndexMap::new());
+
+        Ok(())
+    }
+
+    fn create_table(&self, ident: &Identifier, schema: SchemaRef) -> CatalogResult<TableRef> {
+        let (namespace, table_name) = Self::split_table_ident(ident)?;
+
+        {
+            let tables = self.tables.read().unwrap();
+
+            let Some(namespace_tables) = tables.get(&namespace) else {
+                return Err(CatalogError::ObjectNotFound {
+                    type_: "namespace".to_string(),
+                    ident: namespace.unwrap(),
+                });
+            };
+
+            if namespace_tables.contains_key(table_name) {
+                return Err(CatalogError::obj_already_exists("table", ident));
+            }
+        }
+
+        let table = Arc::new(MemoryTable::new(table_name.to_string(), schema)?);
+
+        self.tables
+            .write()
+            .unwrap()
+            .get_mut(&namespace)
+            .unwrap()
+            .insert(table_name.to_string(), table.clone());
+
+        Ok(table)
+    }
+
+    fn drop_namespace(&self, ident: &Identifier) -> CatalogResult<()> {
+        if ident.has_qualifier() {
+            return Err(CatalogError::obj_not_found("namespace", ident));
+        }
+
+        let namespace = Some(ident.name().to_string());
+
+        match self.tables.write().unwrap().shift_remove(&namespace) {
+            Some(_) => Ok(()),
+            None => Err(CatalogError::obj_not_found("namespace", ident)),
+        }
+    }
+
+    fn drop_table(&self, ident: &Identifier) -> CatalogResult<()> {
+        let (namespace, table_name) = Self::split_table_ident(ident)?;
+
+        let mut tables = self.tables.write().unwrap();
+        let Some(namespace_tables) = tables.get_mut(&namespace) else {
+            return Err(CatalogError::obj_not_found("table", ident));
+        };
+
+        match namespace_tables.shift_remove(table_name) {
+            Some(_) => Ok(()),
+            None => Err(CatalogError::obj_not_found("table", ident)),
+        }
+    }
+
+    fn get_table(&self, ident: &Identifier) -> CatalogResult<TableRef> {
+        let (namespace, table_name) = Self::split_table_ident(ident)?;
+
+        self.tables
+            .read()
+            .unwrap()
+            .get(&namespace)
+            .and_then(|namespace_tables| {
+                namespace_tables
+                    .get(table_name)
+                    .map(|t| t.clone() as TableRef)
+            })
+            .ok_or_else(|| CatalogError::obj_not_found("table", ident))
+    }
+
+    fn has_namespace(&self, ident: &Identifier) -> CatalogResult<bool> {
+        if ident.has_qualifier() {
+            return Ok(false);
+        }
+
+        let namespace = ident.name();
```

Comment:
Let's add a little comment about single-level namespace limitations here since this is a bit subtle.